helix-view/src/register.rs

```diff
@@ -0,0 +1,110 @@
+use crate::document::Document;
+use std::collections::btree_map::Entry;
+use std::collections::BTreeMap;
+use std::fmt;
+
+struct Readonly(String, Box<dyn Fn(&Document) -> Option<String>>);
+
+impl fmt::Debug for Readonly {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_tuple("Readonly").field(&self.0).finish()
+    }
+}
+
+#[derive(Debug)]
+enum Register {
+    Static(Vec<String>),
+    Readonly(Readonly),
+}
+
+/// Currently just wraps a `BTreeMap` of `Register`s
+#[derive(Debug)]
+pub struct Registers {
+    inner: BTreeMap<char, Register>,
+}
+
+impl Registers {
+    pub fn read(&self, doc: &Document, name: char) -> Option<Vec<String>> {
```

Comment:
@the-mikedavis I did what you mentioned but now I don't like the API to always return an owned value, we need to keep cloning.